design/hostclaim-multitenancy-and-hybrid-clusters.md

```diff
@@ -0,0 +1,678 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+<!-- cSpell:ignore Sylva Schiff Kanod argocd GitOps -->
+# HostClaim: multi-tenancy and hybrid clusters
+
+## Status
+
+provisional
+
+## Summary
+
+We introduce a new Custom Resource (named HostClaim) which will facilitate
+the creation of multiple clusters for different tenants. It also provides
+a framework for building clusters with different kind of compute resource:
+baremetal servers but also virtual machines hosted in private or public cloud.
+
+A HostClaim decouples the client need from the actual implementation of the
+compute resource: it establishes a security boundary and provides a way to
+migrate nodes between different kind of compute resources.
+
+A HostClaim expresses that one wants to start a given
+OS image with an initial configuration (typically cloud-init or ignition
+configuration files) on a compute resource that meets a set of requirements
+(host selectors). These requirements can be interpreted either as labels
+for a recyclable resource (such as a bare-metal server) or as characteristics
+for a disposable resource (such as a virtual machine created for the workload).
+The status and meta-data of the HostClaim provide the necessary information
+for the end user to define and manage his workload on the compute resource,
+but they do not grant full control over the resource (typically, BMC
+credentials of servers are not exposed to the tenant).
+
+## Motivation
+
+The standard approach to implementing multi-tenancy in cluster-api is to follow the
+[multi tenancy contract](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/multi-tenancy#contract).
+
+To adhere to this contract with cluster-api-provider-metal3, the clusters must
+be put in different namespaces and the BareMetalHost objects must be defined
+in those namespaces. In this setup, the tenants are the owners of the servers
+and it becomes difficult to share the same server between different clusters if
+they belong to different tenants.
+
+In order to improve server usage, we would like to have a pool of servers that
+clusters can lease depending on their workload. If we maintain the Metal3
+constraints, all clusters must be defined in the same namespace. Unless very
```

Comment:
I will add that this is the BMH namespace. The main difference with the proposals/issues mentioned is that the hostClaim is in the same namespace as the m3m and is owned by it but not the BMH. And it works because BMH are not pivoted.

Reply:
My last comment is wrong in the current version of the proposal. When credentials are used, the hostClaim is not owned by the m3m. This was only true when there was a separate notion of remote hostClaim pointing to the hostClaim where the BMH lives, but we agreed to put the capability to access remote resources in capm3. So the hostClaim are not pivoted.